### DISCUSSION TITLE
ViewModel constructor injection

### DISCUSSION THREAD

**Discussion by roubachof** [PrismLibrary/Prism#3061] [.NET MAUI] [answered]
Since MAUI (or maybe even before?) we can inject dependencies in our view constructor.
If we use the constructor injection strategy to inject our view models, instead of registering the pair `<View, ViewModel>` for navigation, can we just register the `View`, and register the `ViewModel` separately so it can be injected in the View constructor.

From:
```csharp
containerRegistry.RegisterForNavigation<View, ViewModel>();
```

To:
```csharp
containerRegistry.Register<ViewModel>();
containerRegistry.RegisterForNavigation<View>();
```
and:
```csharp
class View {
	public View(ViewModel vm) 
	{
		BindingContext = vm;
	}
}
```

Of course it would be the view responsibility to set the `BindingContext` then.

Would it break something with Prism? 
Normally for what I understood the `Page` is first created, then the `ViewModel` is attached to its `BindingContext`...

**Comment by dansiegel** [comment, marked as answer]
So a couple of things...
1) `RegisterForNavigation` actually does register the ViewModel in Prism.Maui so there is no need to do it yourself. Also it should only be registered as a transient.
2) When you provide the View and ViewModel for Registration it provides a mapping specific to the Route... by default that would be the name of the View. However you could reuse the same View with different ViewModels based on a custom name like `RegisterForNavigation<ViewA, SomeViewModel>()` and `RegisterForNavigation<ViewA, OtherViewModel>("ViewA1")`.
3) Nothing stops you at any point from adding the ViewModel to the Constructor of the View... Prism will detect if the BindingContext is already set and will not try to autowire if you have done that. It's also not a problem because there is a scope at the Page so when the page is created you can still get the correct instance of the NavigationService in the ViewModel injected directly into the Constructor of the View... This is ONLY true for Maui and not Xamarin.Forms. If you are directly injecting the ViewModel into the Constructor of the View you don't need to provide both the View and ViewModel when you call `RegisterForNavigation`.
